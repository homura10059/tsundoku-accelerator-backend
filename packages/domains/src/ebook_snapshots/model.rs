use url::Url;

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}
