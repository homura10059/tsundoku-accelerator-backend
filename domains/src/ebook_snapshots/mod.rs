pub mod repositories;

use crate::infrastructures::prisma;
use crate::infrastructures::prisma::PrismaClient;
use anyhow::Result;
use futures::stream;
use futures::{future, StreamExt};
use headless_chrome::Browser;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = repositories::scraper::get(browser, id.as_str())?;
    let result = repositories::db::insert(client, &snapshot).await?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        let actual = snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
        assert_eq!(actual, ());
    }
}
