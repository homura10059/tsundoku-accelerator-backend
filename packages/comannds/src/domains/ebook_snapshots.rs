pub mod repositories;

use crate::domains::notifications::send_alert_message;
use anyhow::Result;
use db_client::prisma::PrismaClient;
use headless_chrome::Browser;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashSet;
use url::Url;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = match repositories::get(browser, id.as_str()) {
        Ok(s) => s,
        Err(e) => {
            let msg = format!("error snap_ebook id: {},  error:{}", id, e.to_string());
            send_alert_message(msg.clone()).await.unwrap();
            panic!("{}", msg);
        }
    };
    repositories::insert(client, &snapshot).await?;
    Ok(())
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

static PRICE_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"￥(?P<price>\d{1,3}(,\d{3})*)").unwrap());
static POINT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?P<points>\d{1,3}(,\d{3})*)pt").unwrap());

impl Payment {
    fn extract_price(text: &str) -> HashSet<&str> {
        let matched = PRICE_REGEX
            .captures_iter(text)
            .map(|cap| cap.name("price").unwrap().as_str())
            .collect::<Vec<_>>();
        matched.into_iter().collect()
    }

    fn extract_points(text: &str) -> HashSet<&str> {
        let matched = POINT_REGEX
            .captures_iter(text)
            .map(|cap| cap.name("points").unwrap().as_str())
            .collect::<Vec<_>>();
        matched.into_iter().collect()
    }

    pub fn new<T: AsRef<str>>(text: T) -> Option<Self> {
        let price = *Payment::extract_price(text.as_ref()).iter().next()?;
        let points = *Payment::extract_points(text.as_ref()).iter().next()?;
        Some(Payment {
            price: price.to_string().trim().replace(',', ""),
            points: points.to_string().trim().replace(',', ""),
        })
    }
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use db_client::prisma;
    use dotenv;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
    }

    #[test]
    fn test_payment_new() {
        let actual = Payment::new(
            r#"
            Kindle版 (電子書籍)
            ￥3,344
            獲得ポイント: 152pt
            "#,
        );
        let expected = Some(Payment {
            price: String::from("3344"),
            points: String::from("152"),
        });
        assert_eq!(actual, expected);
    }
}
