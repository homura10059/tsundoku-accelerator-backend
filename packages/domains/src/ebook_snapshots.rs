pub mod repositories;

use anyhow::Result;
use headless_chrome::Browser;
use infrastructures::prisma::PrismaClient;
use url::Url;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = repositories::scraper::get(browser, id.as_str())?;
    repositories::db::insert(client, &snapshot).await?;
    Ok(())
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct Payment {
    pub price: String,
    pub points: String,
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct EbookSnapshot {
    pub ebook_id: String,
    pub scraped_at: i64,
    pub thumbnail_url: Url,
    pub payment_ebook: Option<Payment>,
    pub payment_real: Option<Payment>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;
    use infrastructures::prisma;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        let actual = snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
        assert_eq!(actual, ());
    }
}
