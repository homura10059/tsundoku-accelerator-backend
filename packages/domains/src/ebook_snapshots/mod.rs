mod model;
pub mod repositories;

use anyhow::Result;
use headless_chrome::Browser;
use infrastructures::prisma::PrismaClient;

pub async fn snap_ebook(client: &PrismaClient, browser: &Browser, id: String) -> Result<()> {
    let snapshot = repositories::scraper::get(browser, id.as_str())?;
    repositories::db::insert(client, &snapshot).await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;
    use infrastructures::prisma;

    #[tokio::test]
    async fn it_works_snap_ebook() {
        dotenv::dotenv().ok();
        let client = prisma::new_client().await.unwrap();
        let browser = Browser::default().unwrap();

        let actual = snap_ebook(&client, &browser, String::from("B00XV8YCJI"))
            .await
            .unwrap();
        assert_eq!(actual, ());
    }
}
