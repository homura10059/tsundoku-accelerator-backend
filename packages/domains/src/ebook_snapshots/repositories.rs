pub mod db;
pub mod scraper;
