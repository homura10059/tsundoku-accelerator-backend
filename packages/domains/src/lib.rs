pub mod ebook_snapshots;
pub mod ebooks;
mod infrastructures;
pub mod item_metadata;
mod models;
pub mod notifications;
pub mod wish_lists;
