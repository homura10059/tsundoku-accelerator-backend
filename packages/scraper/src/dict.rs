use std::collections::HashMap;

pub fn from(attr: &[String]) -> HashMap<String, String> {
    attr.chunks(2)
        .filter_map(|chunk| {
            let key = chunk.get(0)?;
            let val = chunk.get(1)?;
            Some((key.to_string(), val.to_string()))
        })
        .collect::<HashMap<String, String>>()
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    proptest! {
        #[test]
        fn doesnt_crash(attr: Vec<String>) {
            from(&attr);
        }
    }
}
