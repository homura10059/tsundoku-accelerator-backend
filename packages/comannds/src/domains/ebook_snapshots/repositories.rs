use crate::domains::ebook_snapshots::{EbookSnapshot, Payment};
use anyhow::anyhow;
use chrono::Utc;
use db_client::prisma::{ebook, ebook_snapshot, PrismaClient};
use headless_chrome::Browser;
use math::round;
use scraper::dict::from;
use url::Url;

pub fn create_url(id: &str) -> anyhow::Result<Url> {
    let url = Url::parse("https://www.amazon.co.jp/dp/")?;
    let joined = url.join(id)?;
    Ok(joined)
}

pub fn get(browser: &Browser, id: &str) -> anyhow::Result<EbookSnapshot> {
    let url = create_url(id)?;

    let tab = browser.new_tab()?;
    tab.navigate_to(url.as_str())?;
    tab.wait_for_element("#navFooter")?;

    let image = tab.find_element("#ebooksImgBlkFront")?;
    let image_attribute = image.get_attributes()?.unwrap();
    let dict = from(&image_attribute);
    let thumbnail_url_str = dict.get("src").unwrap();
    let thumbnail_url = Url::parse(thumbnail_url_str.as_str())?;

    let price_list = tab.find_elements("#tmmSwatches .a-button-text")?;
    let payments = price_list
        .iter()
        .flat_map(|price| price.get_inner_text())
        .flat_map(Payment::new)
        .collect::<Vec<_>>();

    let snapshot = EbookSnapshot {
        ebook_id: id.to_string(),
        scraped_at: Utc::now().timestamp(),
        thumbnail_url,
        payment_ebook: payments.get(0).cloned(),
        payment_real: payments.get(1).cloned(),
    };

    tab.close(true)?;

    Ok(snapshot)
}

pub async fn insert(client: &PrismaClient, ebook_snapshot: &EbookSnapshot) -> anyhow::Result<()> {
    let payment_ebook = ebook_snapshot.payment_ebook.clone().ok_or(anyhow!(
        "missing payment_ebook in id:{}",
        ebook_snapshot.ebook_id
    ))?;

    let price = payment_ebook.price.parse::<f64>()?;
    let points = payment_ebook.points.parse::<f64>()?;

    let points_rate = round::floor(points / price * 100.0, 2);

    let payment_real = ebook_snapshot.payment_real.clone();

    let real_price = payment_real.and_then(|payment| payment.price.parse::<f64>().ok());
    let discount = real_price.map(|real| real - price);
    let discount_rate = discount.map(|dis| round::floor(dis / price * 100.0, 2));

    client
        .ebook_snapshot()
        .create(
            ebook::UniqueWhereParam::IdEquals(ebook_snapshot.ebook_id.clone()),
            ebook_snapshot.scraped_at,
            ebook_snapshot.thumbnail_url.to_string(),
            price,
            points,
            points_rate,
            vec![
                ebook_snapshot::SetParam::SetDiscount(discount),
                ebook_snapshot::SetParam::SetDiscountRate(discount_rate),
            ],
        )
        .exec()
        .await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domains::ebook_snapshots::{EbookSnapshot, Payment};
    use db_client::prisma;
    use dotenv;
    use insta::assert_debug_snapshot;
    use url::Url;

    #[tokio::test]
    async fn test_insert() {
        dotenv::dotenv().ok();

        let payment_ebook = Payment {
            price: String::from("1000"),
            points: String::from("11"),
        };

        let payment_real = Payment {
            price: String::from("1100"),
            points: String::from("11"),
        };

        let expected = EbookSnapshot {
            ebook_id: String::from("B09RQGMYKZ"),
            scraped_at: 0,
            thumbnail_url: Url::parse("https://m.media-amazon.com/images/I/51CTnaTcJtL.jpg")
                .unwrap(),
            payment_ebook: Some(payment_ebook),
            payment_real: Some(payment_real),
        };

        let client = prisma::new_client().await.unwrap();

        insert(&client, &expected).await.unwrap()
    }

    #[test]
    fn test_get() {
        let browser = Browser::default().unwrap();
        let id = String::from("B09RQGMYKZ");
        let actual = get(&browser, id.as_str()).unwrap();
        assert_eq!(actual.ebook_id, id);
        assert_debug_snapshot!(actual.payment_ebook);
        assert_debug_snapshot!(actual.payment_real);
    }
}
