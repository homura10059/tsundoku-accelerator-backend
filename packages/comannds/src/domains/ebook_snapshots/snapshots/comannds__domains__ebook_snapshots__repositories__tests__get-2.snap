---
source: packages/comannds/src/domains/ebook_snapshots/repositories.rs
expression: actual.payment_real
---
Some(
    Payment {
        price: "3520",
        points: "106",
    },
)
